File: backend/app/services/calculator.py
"""
calculator.py — Core P&L computation engine.

Refactored from scripts.py.  Implements both the original v1 flat-cost model
and the v2 making_cost + packaging_cost model described in Readme.md §6.

Key design decisions preserved from the original:
- One final status per Sub Order No, resolved by STATUS_PRIORITY.
- COGS charged based on resolved status × configurable loss rates.
- Ads/Referral/Compensation are account-level only — never allocated per-SKU.
"""

from __future__ import annotations

from typing import IO

import pandas as pd

from app.config import (
    COGS_FULL_STATUSES,
    COGS_ZERO_STATUSES,
    CONFIGURABLE_LOSS_STATUSES,
    STATUS_PRIORITY,
)
from app.services.parser import _to_num, parse_small_sheet


# ---------------------------------------------------------------------------
# Step 1: Collapse multi-row settlement legs into one row per order
# ---------------------------------------------------------------------------

def resolve_order_level(df: pd.DataFrame) -> pd.DataFrame:
    """Collapse the Order Payments DataFrame to one row per Sub Order No.

    For each order:
    - Settlement amounts are SUMMED (handles multi-leg payments).
    - Status is resolved to the highest-priority terminal outcome.
    - SKU, product name, order date come from the first row.
    - Quantity takes the MAX (same line item, not summed across legs).
    """

    def _pick_status(statuses: pd.Series) -> str | None:
        clean = [s for s in statuses if pd.notna(s)]
        if not clean:
            return None
        return max(clean, key=lambda s: STATUS_PRIORITY.get(s.lower(), 0))

    grouped = df.groupby("Sub Order No").agg(
        supplier_sku=("Supplier SKU", "first"),
        product_name=("Product Name", "first") if "Product Name" in df.columns else ("Supplier SKU", "first"),
        order_date=("Order Date", "min"),
        payment_date=("Payment Date", "max"),
        quantity=("Quantity", "max"),
        gross_sale_amount=("Total Sale Amount (Incl. Shipping & GST)", "sum"),
        net_settlement=("Final Settlement Amount", "sum"),
        status=("Live Order Status", _pick_status),
    ).reset_index()

    grouped["status_norm"] = grouped["status"].fillna("unknown").str.lower()
    return grouped


# ---------------------------------------------------------------------------
# Step 2: Apply COGS (v2 model: making + packaging split)
# ---------------------------------------------------------------------------

def apply_cogs(
    order_level: pd.DataFrame,
    sku_costs: dict[str, dict],
    loss_rates: dict[str, float],
) -> pd.DataFrame:
    """Apply v2 COGS logic with making_cost + packaging_cost split.

    sku_costs: {sku: {"making_cost": float, "packaging_cost": float}}
    loss_rates: {"rto": float, "return": float, "lost": float, "unresolved": float}

    v2 COGS rules (from Readme.md §6):
        Delivered / Exchange → making + packaging (100%)
        RTO                 → packaging only (making cost NOT charged)
        Customer Return     → making + packaging (100%)
        Lost                → making + packaging (100%)
        Cancelled           → nothing
        Unresolved/Shipped  → uses unresolved loss rate × (making + packaging)
    """

    def _lookup_cost(sku: str, component: str) -> float:
        entry = sku_costs.get(sku)
        if entry is None:
            return 0.0
        return entry.get(component, 0.0)

    order_level["making_cost"] = order_level["supplier_sku"].apply(
        lambda s: _lookup_cost(s, "making_cost")
    )
    order_level["packaging_cost"] = order_level["supplier_sku"].apply(
        lambda s: _lookup_cost(s, "packaging_cost")
    )
    order_level["unit_cost"] = order_level["making_cost"] + order_level["packaging_cost"]
    order_level["cost_mapped"] = order_level["supplier_sku"].apply(
        lambda s: s in sku_costs and sku_costs[s] is not None
    )

    def _compute_cogs(row: pd.Series) -> tuple[float, float]:
        """Return (making_cogs, packaging_cogs) for one order."""
        status = row["status_norm"]
        qty = row["quantity"]
        making = row["making_cost"]
        packaging = row["packaging_cost"]

        if status in COGS_ZERO_STATUSES:
            return 0.0, 0.0

        if status in COGS_FULL_STATUSES:
            return making * qty, packaging * qty

        pkg_rate = loss_rates.get("packaging_loss", 1.0)

        if status == "rto":
            # RTO: product comes back resellable
            rate = loss_rates.get("rto", 0.0)
            return making * qty * rate, packaging * qty * pkg_rate

        if status == "return":
            rate = loss_rates.get("return", 1.0)
            return making * qty * rate, packaging * qty * pkg_rate

        if status == "lost":
            rate = loss_rates.get("lost", 1.0)
            return making * qty * rate, packaging * qty * pkg_rate

        # Unresolved / shipped / unknown
        rate = loss_rates.get("unresolved", 0.0)
        return making * qty * rate, packaging * qty * pkg_rate

    cogs_split = order_level.apply(_compute_cogs, axis=1, result_type="expand")
    cogs_split.columns = ["cogs_making", "cogs_packaging"]

    order_level["cogs_making"] = cogs_split["cogs_making"]
    order_level["cogs_packaging"] = cogs_split["cogs_packaging"]
    order_level["cogs"] = order_level["cogs_making"] + order_level["cogs_packaging"]
    order_level["profit"] = order_level["net_settlement"] - order_level["cogs"]

    return order_level


# ---------------------------------------------------------------------------
# Step 3: Summarise account-level adjustments (ads, referral, comp/recovery)
# ---------------------------------------------------------------------------

def summarize_account_level(source) -> dict[str, float]:
    """Extract ads cost, referral income, and compensation/recovery totals.

    These are account-level (not per-SKU) because Meesho's export provides
    no SKU or order linkage for these sheets.
    """
    out = {"ads_cost": 0.0, "referral_income": 0.0, "compensation_recovery": 0.0}

    ads = parse_small_sheet(source, "Ads Cost")
    if not ads.empty and "Total Ads Cost" in ads.columns:
        out["ads_cost"] = _to_num(ads["Total Ads Cost"]).sum()

    # Need to re-open the file for each sheet if source is a file-like object
    ref = parse_small_sheet(source, "Referral Payments")
    if not ref.empty and "Net Referral Amount" in ref.columns:
        out["referral_income"] = _to_num(ref["Net Referral Amount"]).sum()

    comp = parse_small_sheet(source, "Compensation and Recovery")
    if not comp.empty and "Amount (inc GST) INR" in comp.columns:
        out["compensation_recovery"] = _to_num(comp["Amount (inc GST) INR"]).sum()

    return out


# ---------------------------------------------------------------------------
# Step 4: Build the SKU-wise P&L report
# ---------------------------------------------------------------------------

def build_sku_report(order_level: pd.DataFrame) -> pd.DataFrame:
    """Aggregate order-level data into a SKU-wise P&L table, sorted by profit."""

    sku = order_level.groupby("supplier_sku").agg(
        orders=("Sub Order No", "count"),
        units=("quantity", "sum"),
        gross_sale_amount=("gross_sale_amount", "sum"),
        net_settlement=("net_settlement", "sum"),
        cogs=("cogs", "sum"),
        cogs_making=("cogs_making", "sum"),
        cogs_packaging=("cogs_packaging", "sum"),
        profit=("profit", "sum"),
        cost_mapped=("cost_mapped", "all"),
        product_name=("product_name", "first"),
    ).reset_index().rename(columns={"supplier_sku": "sku"})

    # Margin % only meaningful when net settlement is positive
    positive_rev = sku["net_settlement"] > 0
    sku["margin_pct"] = None
    sku.loc[positive_rev, "margin_pct"] = (
        sku.loc[positive_rev, "profit"]
        / sku.loc[positive_rev, "net_settlement"]
        * 100
    ).round(2)

    sku = sku.sort_values("profit", ascending=False)
    return sku


# ---------------------------------------------------------------------------
# Step 5: Identify pending (unsettled) orders
# ---------------------------------------------------------------------------

def find_pending_orders(
    orders_df: pd.DataFrame, settled_sub_order_nos: set[str]
) -> pd.DataFrame:
    """Find orders placed in the window that haven't been settled yet."""
    return orders_df[~orders_df["Sub Order No"].isin(settled_sub_order_nos)].copy()


# ---------------------------------------------------------------------------
# Step 6: Build the status breakdown for the donut chart
# ---------------------------------------------------------------------------

def build_status_breakdown(order_level: pd.DataFrame) -> list[dict]:
    """Aggregate orders by resolved status for the status-donut chart."""
    total = len(order_level)
    breakdown = (
        order_level.groupby("status_norm")
        .agg(
            order_count=("Sub Order No", "count"),
            total_settlement=("net_settlement", "sum"),
        )
        .reset_index()
        .rename(columns={"status_norm": "status"})
    )
    breakdown["percentage"] = (breakdown["order_count"] / total * 100).round(2)
    return breakdown.to_dict(orient="records")

File: frontend/src/services/calculator.js
// frontend/src/services/calculator.js
import {
  STATUS_PRIORITY,
  COGS_FULL_STATUSES,
  COGS_ZERO_STATUSES
} from "./config";
import { parseOrderPayments, parseSmallSheet, parseOrdersCsv } from "./parser";
import { getUnmappedSkus, loadCosts } from "./skuCosts";

/**
 * Collapse the Order Payments array to one row per Sub Order No.
 */
export const resolveOrderLevel = (df) => {
  const grouped = new Map();

  df.forEach(row => {
    const subOrderNo = row["Sub Order No"];
    if (!grouped.has(subOrderNo)) {
      grouped.set(subOrderNo, {
        "Sub Order No": subOrderNo,
        supplier_sku: row["Supplier SKU"],
        product_name: row["Product Name"] || row["Supplier SKU"],
        order_date: row["Order Date"],
        payment_date: row["Payment Date"],
        quantity: row["Quantity"],
        gross_sale_amount: 0,
        net_settlement: 0,
        return_shipping_charge: 0,
        statuses: []
      });
    }

    const group = grouped.get(subOrderNo);

    // Update min order_date
    if (row["Order Date"] && (!group.order_date || row["Order Date"] < group.order_date)) {
      group.order_date = row["Order Date"];
    }

    // Update max payment_date
    if (row["Payment Date"] && (!group.payment_date || row["Payment Date"] > group.payment_date)) {
      group.payment_date = row["Payment Date"];
    }

    // Quantity takes MAX
    if (row["Quantity"] > group.quantity) {
      group.quantity = row["Quantity"];
    }

    // Summing settlements
    group.gross_sale_amount += row["Total Sale Amount (Incl. Shipping & GST)"] || 0;
    group.net_settlement += row["Final Settlement Amount"] || 0;
    group.return_shipping_charge += row["Return Shipping Charge"] || 0;

    if (row["Live Order Status"] !== null && row["Live Order Status"] !== undefined) {
      group.statuses.push(row["Live Order Status"]);
    }
  });

  const result = Array.from(grouped.values()).map(group => {
    // resolve status
    let finalStatus = null;
    let maxPriority = -1;

    group.statuses.forEach(s => {
      const priority = STATUS_PRIORITY[s.toLowerCase()] || 0;
      if (priority > maxPriority) {
        maxPriority = priority;
        finalStatus = s;
      }
    });

    group.status = finalStatus;
    group.status_norm = finalStatus ? finalStatus.toLowerCase() : "unknown";
    delete group.statuses;

    return group;
  });

  return result;
};

/**
 * Apply COGS
 */
export const applyCogs = (orderLevel, skuCosts, lossRates) => {
  return orderLevel.map(row => {
    const sku = row.supplier_sku;
    const costEntry = skuCosts[sku] || { making_cost: 0, packaging_cost: 0 };
    const makingCost = costEntry.making_cost || 0;
    const packagingCost = costEntry.packaging_cost || 0;

    row.making_cost = makingCost;
    row.packaging_cost = packagingCost;
    row.unit_cost = makingCost + packagingCost;
    row.cost_mapped = (sku in skuCosts) && (skuCosts[sku] !== null);

    let cogsMaking = 0;
    let cogsPackaging = 0;

    const status = row.status_norm;
    const qty = row.quantity;

    if (COGS_ZERO_STATUSES.has(status)) {
      cogsMaking = 0;
      cogsPackaging = 0;
    } else if (COGS_FULL_STATUSES.has(status)) {
      cogsMaking = makingCost * qty;
      cogsPackaging = packagingCost * qty;
    } else if (status === "rto") {
      cogsMaking = makingCost * qty * (lossRates.rto || 0.0);
      cogsPackaging = packagingCost * qty * (lossRates.rto_packaging_loss !== undefined ? lossRates.rto_packaging_loss : 1.0);
    } else if (status === "return") {
      cogsMaking = makingCost * qty * (lossRates.return !== undefined ? lossRates.return : 1.0);
      cogsPackaging = packagingCost * qty * (lossRates.return_packaging_loss !== undefined ? lossRates.return_packaging_loss : 1.0);
    } else if (status === "lost") {
      cogsMaking = makingCost * qty * (lossRates.lost !== undefined ? lossRates.lost : 1.0);
      cogsPackaging = packagingCost * qty * 1.0;
    } else {
      // Unresolved / shipped / unknown
      cogsMaking = makingCost * qty * (lossRates.unresolved !== undefined ? lossRates.unresolved : 0.0);
      cogsPackaging = packagingCost * qty * 1.0;
    }

    row.cogs_making = cogsMaking;
    row.cogs_packaging = cogsPackaging;
    row.cogs = cogsMaking + cogsPackaging;
    row.profit = row.net_settlement - row.cogs;

    return row;
  });
};

/**
 * Summarise account-level adjustments
 */
export const summarizeAccountLevel = (workbook) => {
  const out = { ads_cost: 0.0, referral_income: 0.0, compensation_recovery: 0.0 };

  const ads = parseSmallSheet(workbook, "Ads Cost");
  ads.forEach(r => {
    out.ads_cost += Number(r["Total Ads Cost"] || 0);
  });

  const ref = parseSmallSheet(workbook, "Referral Payments");
  ref.forEach(r => {
    out.referral_income += Number(r["Net Referral Amount"] || 0);
  });

  const comp = parseSmallSheet(workbook, "Compensation and Recovery");
  comp.forEach(r => {
    out.compensation_recovery += Number(r["Amount (inc GST) INR"] || 0);
  });

  return out;
};

/**
 * Build SKU report
 */
export const buildSkuReport = (orderLevel) => {
  const skuMap = new Map();

  orderLevel.forEach(row => {
    const sku = row.supplier_sku;
    if (!skuMap.has(sku)) {
      skuMap.set(sku, {
        sku: sku,
        product_name: row.product_name,
        orders: 0,
        units: 0,
        delivered_orders: 0,
        rto_orders: 0,
        return_orders: 0,
        gross_sale_amount: 0,
        net_settlement: 0,
        cogs: 0,
        cogs_making: 0,
        cogs_packaging: 0,
        rto_cost: 0,
        profit: 0,
        cost_mapped: true
      });
    }

    const s = skuMap.get(sku);
    s.orders += 1;
    s.units += row.quantity;
    if (row.status_norm === "delivered") s.delivered_orders += 1;
    if (row.status_norm === "rto") {
      s.rto_orders += 1;
      s.rto_cost += row.cogs;
    }
    if (row.status_norm === "return") s.return_orders += 1;

    s.gross_sale_amount += row.gross_sale_amount;
    s.net_settlement += row.net_settlement;
    s.cogs += row.cogs;
    s.cogs_making += row.cogs_making;
    s.cogs_packaging += row.cogs_packaging;
    s.profit += row.profit;
    if (!row.cost_mapped) {
      s.cost_mapped = false;
    }
  });

  const result = Array.from(skuMap.values()).map(s => {
    if (s.net_settlement > 0) {
      s.margin_pct = (s.profit / s.net_settlement) * 100;
    } else {
      s.margin_pct = null;
    }
    s.avg_sale_price = s.units > 0 ? (s.gross_sale_amount / s.units) : 0;
    return s;
  });

  result.sort((a, b) => b.profit - a.profit);
  return result;
};

/**
 * Find pending orders
 */
export const findPendingOrders = (ordersDf, settledSubOrderNos) => {
  return ordersDf.filter(row => !settledSubOrderNos.has(row["Sub Order No"]));
};

/**
 * Build status breakdown
 */
export const buildStatusBreakdown = (orderLevel) => {
  const map = new Map();
  let total = orderLevel.length;

  orderLevel.forEach(r => {
    const s = r.status_norm;
    if (!map.has(s)) {
      map.set(s, { status: s, order_count: 0, total_settlement: 0 });
    }
    const grp = map.get(s);
    grp.order_count += 1;
    grp.total_settlement += r.net_settlement;
  });

  return Array.from(map.values()).map(grp => {
    grp.percentage = total > 0 ? (grp.order_count / total * 100) : 0;
    return grp;
  });
};

/**
 * Safe round function similar to backend
 */
const safeRound = (val) => {
  if (val === null || val === undefined || isNaN(val)) return 0;
  return Math.round(val * 100) / 100;
};

/**
 * Orchestrator: Compute PnL directly in the browser
 */
export const computePnl = async (workbook, ordersDf, lossRatesReq) => {
  // 1. Parse order payments
  const orderPaymentsRaw = parseOrderPayments(workbook);

  // Load SKU costs from localStorage
  const skuCosts = loadCosts();

  const lossRates = {
    rto: lossRatesReq.rto,
    return: lossRatesReq.return_rate,
    lost: lossRatesReq.lost,
    unresolved: lossRatesReq.unresolved,
    rto_packaging_loss: lossRatesReq.rto_packaging_loss,
    return_packaging_loss: lossRatesReq.return_packaging_loss,
  };

  // 2. Resolve to one row per sub order no
  let orderLevel = resolveOrderLevel(orderPaymentsRaw);

  // 3. Apply COGS
  orderLevel = applyCogs(orderLevel, skuCosts, lossRates);

  // 4. Account-level adjustments
  const extras = summarizeAccountLevel(workbook);

  // 5. Build overall PnL
  const net_settlement = orderLevel.reduce((acc, row) => acc + row.net_settlement, 0);
  const return_shipping_charge = orderLevel.reduce((acc, row) => acc + (row.return_shipping_charge || 0), 0);
  const cogs = orderLevel.reduce((acc, row) => acc + row.cogs, 0);
  const cogs_making = orderLevel.reduce((acc, row) => acc + row.cogs_making, 0);
  const cogs_packaging = orderLevel.reduce((acc, row) => acc + row.cogs_packaging, 0);

  const cogs_making_lost = orderLevel.reduce((acc, row) => acc + (!COGS_FULL_STATUSES.has(row.status_norm) ? row.cogs_making : 0), 0);
  const cogs_packaging_lost = orderLevel.reduce((acc, row) => acc + (!COGS_FULL_STATUSES.has(row.status_norm) ? row.cogs_packaging : 0), 0);

  const making_loss_rto = orderLevel.reduce((acc, row) => acc + (row.status_norm === 'rto' ? row.cogs_making : 0), 0);
  const making_loss_return = orderLevel.reduce((acc, row) => acc + (row.status_norm === 'return' ? row.cogs_making : 0), 0);
  const packaging_loss_rto = orderLevel.reduce((acc, row) => acc + (row.status_norm === 'rto' ? row.cogs_packaging : 0), 0);
  const packaging_loss_return = orderLevel.reduce((acc, row) => acc + (row.status_norm === 'return' ? row.cogs_packaging : 0), 0);
  const gross_profit = net_settlement - cogs;
  const net_profit = gross_profit + extras.ads_cost + extras.referral_income + extras.compensation_recovery;

  // dates
  let payment_start = null;
  let payment_end = null;
  orderPaymentsRaw.forEach(r => {
    if (r["Payment Date"]) {
      if (!payment_start || r["Payment Date"] < payment_start) payment_start = r["Payment Date"];
      if (!payment_end || r["Payment Date"] > payment_end) payment_end = r["Payment Date"];
    }
  });

  const total_units = orderLevel.reduce((acc, row) => acc + row.quantity, 0);

  const formatDate = (dateObj) => {
    if (!dateObj) return "";
    const y = dateObj.getFullYear();
    const m = String(dateObj.getMonth() + 1).padStart(2, '0');
    const d = String(dateObj.getDate()).padStart(2, '0');
    return `${y}-${m}-${d}`;
  };

  const overall = {
    net_settlement: safeRound(net_settlement),
    cogs: safeRound(cogs),
    cogs_making: safeRound(cogs_making),
    cogs_packaging: safeRound(cogs_packaging),
    cogs_making_lost: safeRound(cogs_making_lost),
    cogs_packaging_lost: safeRound(cogs_packaging_lost),
    making_loss_rto: safeRound(making_loss_rto),
    making_loss_return: safeRound(making_loss_return),
    packaging_loss_rto: safeRound(packaging_loss_rto),
    packaging_loss_return: safeRound(packaging_loss_return),
    return_shipping_charge: safeRound(return_shipping_charge),
    gross_profit: safeRound(gross_profit),
    ads_cost: safeRound(extras.ads_cost),
    referral_income: safeRound(extras.referral_income),
    compensation_recovery: safeRound(extras.compensation_recovery),
    net_profit: safeRound(net_profit),
    payment_window_start: formatDate(payment_start),
    payment_window_end: formatDate(payment_end),
    total_orders: orderLevel.length,
    total_units: safeRound(total_units)
  };

  // 6. SKU-wise report
  const sku_df = buildSkuReport(orderLevel);
  const sku_rows = sku_df.map(row => ({
    ...row,
    units: safeRound(row.units),
    gross_sale_amount: safeRound(row.gross_sale_amount),
    net_settlement: safeRound(row.net_settlement),
    cogs: safeRound(row.cogs),
    cogs_making: safeRound(row.cogs_making),
    cogs_packaging: safeRound(row.cogs_packaging),
    rto_cost: safeRound(row.rto_cost),
    profit: safeRound(row.profit),
    margin_pct: row.margin_pct !== null ? safeRound(row.margin_pct) : null,
    avg_sale_price: safeRound(row.avg_sale_price)
  }));

  // 7. Status breakdown
  const status_breakdown = buildStatusBreakdown(orderLevel).map(item => ({
    ...item,
    total_settlement: safeRound(item.total_settlement),
    percentage: safeRound(item.percentage)
  }));

  // 8. Unmapped SKUs
  const allSkus = new Set(orderPaymentsRaw.map(r => r["Supplier SKU"]).filter(Boolean));
  const unmapped = getUnmappedSkus(allSkus, skuCosts);
  const unmapped_list = Array.from(unmapped).sort().map(sku => {
    const skuOrders = orderLevel.filter(r => r.supplier_sku === sku);
    const setAmt = skuOrders.reduce((acc, r) => acc + r.net_settlement, 0);
    return {
      sku,
      order_count: skuOrders.length,
      settlement_amount: safeRound(setAmt)
    };
  });

  // 9. Review orders
  const review_orders = orderLevel.filter(r => !r.status).map(r => ({
    sub_order_no: r["Sub Order No"],
    sku: r.supplier_sku,
    net_settlement: safeRound(r.net_settlement)
  }));

  // 10. Pending orders
  let pending = null;
  if (ordersDf && ordersDf.length > 0) {
    const settledIds = new Set(orderLevel.map(r => r["Sub Order No"]));
    const pendingDf = findPendingOrders(ordersDf, settledIds);

    if (pendingDf.length > 0) {
      const byStatusMap = new Map();
      pendingDf.forEach(r => {
        const status = r["Reason for Credit Entry"] || "Unknown";
        if (!byStatusMap.has(status)) {
          byStatusMap.set(status, { reason: status, orders: 0, gross_value: 0 });
        }
        const st = byStatusMap.get(status);
        st.orders += 1;
        st.gross_value += r["Supplier Discounted Price (Incl GST and Commision)"] || 0;
      });

      const totalPendingGross = pendingDf.reduce((acc, r) => acc + (r["Supplier Discounted Price (Incl GST and Commision)"] || 0), 0);

      pending = {
        total_pending: pendingDf.length,
        total_gross_value: safeRound(totalPendingGross),
        by_status: Array.from(byStatusMap.values()).map(s => ({
          ...s,
          gross_value: safeRound(s.gross_value)
        }))
      };
    } else {
      pending = { total_pending: 0, total_gross_value: 0, by_status: [] };
    }
  }

  return {
    overall,
    sku_rows,
    status_breakdown,
    unmapped_skus: unmapped_list,
    pending_orders: pending,
    review_orders,
    loss_rates: lossRatesReq
  };
};
